solr/core/src/test/org/apache/solr/cloud/overseer/ZkStateWriterTest.java

```diff
@@ -302,6 +297,80 @@ public void testSingleExternalCollection() throws Exception {
     }
 
 
+  }
+
+  public void testSingleExternalCollectionCompressedState() throws Exception {
+    Path zkDir = createTempDir("testSingleExternalCollection");
+
+    ZkTestServer server = new ZkTestServer(zkDir);
+
+    SolrZkClient zkClient = null;
+
+    try {
+      server.run();
+
+      zkClient = new SolrZkClient(server.getZkAddress(), OverseerTest.DEFAULT_CONNECTION_TIMEOUT);
+      ZkController.createClusterZkNodes(zkClient);
+
+      try (ZkStateReader reader = new ZkStateReader(zkClient)) {
+        reader.createClusterStateWatchersAndUpdate();
+
+        ZkStateWriter writer = new ZkStateWriter(reader, new Stats(), 500000);
+
+        zkClient.makePath(ZkStateReader.COLLECTIONS_ZKNODE + "/c1", true);
+
+        // create new collection with stateFormat = 2
+        ZkWriteCommand c1 = new ZkWriteCommand("c1",
+                new DocCollection("c1", new HashMap<String, Slice>(), new HashMap<String, Object>(), DocRouter.DEFAULT, 0, ZkStateReader.COLLECTIONS_ZKNODE + "/c1/state.json"));
+
+        writer.enqueueUpdate(reader.getClusterState(), Collections.singletonList(c1), null);
+        writer.writePendingUpdates();
+
+        Map map = (Map) Utils.fromJSON(zkClient.getData("/clusterstate.json", null, null, true));
+        assertNull(map.get("c1"));
+        byte[] data = zkClient.getData(ZkStateReader.COLLECTIONS_ZKNODE + "/c1/state.json", null, null, true);
+        map = (Map) Utils.fromJSON(data);
+        assertNotNull(map.get("c1"));
+      }
+
+      try (ZkStateReader reader = new ZkStateReader(zkClient)) {
+        reader.createClusterStateWatchersAndUpdate();
+
+        ZkStateWriter writer = new ZkStateWriter(reader, new Stats(), 500000);
+
+        zkClient.makePath(ZkStateReader.COLLECTIONS_ZKNODE + "/c2", true);
+
+        // create new collection with stateFormat = 2
+        Map<String, Slice> slices = new HashMap<>();
+        for (int i = 0; i < 4096; i++) {
+          Map<String, Replica> replicas = new HashMap<>();
+          Map<String, Object> replicaProps = new HashMap<>();
+          replicaProps.put(ZkStateReader.NODE_NAME_PROP, "node1:8983_8983");
+          replicaProps.put(ZkStateReader.CORE_NAME_PROP, "coreNode" + i);
+          replicaProps.put(ZkStateReader.REPLICA_TYPE, "NRT");
+          replicas.put("coreNode" + i, new Replica("coreNode" + i, replicaProps, "c2", "shard" + i));
+          slices.put("shard" + i, new Slice("shard" + i, replicas, new HashMap<>(), "c2"));
+        }
```

Comment:
nit: the assumption here is that we're creating a `state.json` file that will meet/exceed the threshold for compression? Would make sense to add a comment documenting that assumption/intention (and would it be possible to also read directly from Zk, bypassing zkClient, to ensure that the data is really compressed?).